### PR TITLE
ALTAPPS-1041: Shared study plan screen optimization

### PR DESCRIPTION
**Pull request by ivan-magda**
**YouTrack Issues**:
[#ALTAPPS-1041](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1041)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**

1. Adds `/api/learning-activities/with-sections` support
2. Replaces initial data loading flow with new endpoint

_We have a lot of features that depends on current study plan state, so I leaved fetching of the current study plan in background on every screen appear like we did before, otherwise this introduces bugs._

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetFeature.kt [hunk 1]
```diff
@@ -82,20 +73,18 @@ object StudyPlanWidgetFeature {
...
-    }
+        object ReloadContentInBackground : InternalMessage
 
-    internal sealed interface SectionsFetchResult : Message {
-        data class Success(val sections: List<StudyPlanSection>) : SectionsFetchResult
+        object FetchLearningActivitiesWithSectionsFailed : InternalMessage
+        data class FetchLearningActivitiesWithSectionsSuccess(
+            val learningActivities: List<LearningActivity>,
+            val studyPlanSections: List<StudyPlanSection>
+        ) : InternalMessage
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetFeature.kt:85]: I suggest to make it a sealed class, like `LearningActivitiesFetchResult` to make all messages uniform

**Review by XanderZhu — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/learning_activities/data/repository/LearningActivitiesRepositoryImpl.kt [hunk 2]
```diff
@@ -46,4 +49,17 @@ class LearningActivitiesRepositoryImpl(
                 types = types
             )
         )
+
+    override suspend fun getLearningActivitiesWithSections(
+        studyPlanSectionTypes: Set<StudyPlanSectionType>,
+        learningActivityTypes: Set<LearningActivityType>,
+        learningActivityStates: Set<LearningActivityState>
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/learning_activities/data/repository/LearningActivitiesRepositoryImpl.kt:56]: As I remember, we don't have such filtering in new API, the endpoint always returns only TODO activities

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt [hunk 3]
```diff
@@ -114,61 +101,39 @@ class StudyPlanWidgetReducer : StateReducer<State, Message, Action> {
...
     ): StudyPlanWidgetReducerResult {
-        if (state.studyPlan == null) {
-            return state to emptySet()
-        }
-
-        val sortedSupportedSections = message.sections
-            .filter { it.isVisible && StudyPlanSectionType.supportedTypes().contains(it.type) }
-            .sortedBy { state.studyPlan.sections.indexOf(it.id) }
+        val visibleSections = message.studyPlanSections.filter { it.isVisible }
+        val currentSectionId = visibleSections.firstOrNull()?.id
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt:126]: We should determine current section in other way. Current section is the section that contains learning activities that was returned by that API. In current realization there could be situation when API returned activities from not first visible section. In this way we should hide all visible section that above of current section. For example, the first visible section contains only not supported activities.

**Review by vladkash — CHANGES_REQUESTED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt (hunk 3 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt:126]: OMG, why like that 🤯

File: shared/src/commonMain/kotlin/org/hyperskill/app/learning_activities/data/repository/LearningActivitiesRepositoryImpl.kt (hunk 2 above)
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/learning_activities/data/repository/LearningActivitiesRepositoryImpl.kt:56]: We have such filtering in the new API

File: shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt (hunk 3 above)
Comment (vladkash) [reply shared/src/commonMain/kotlin/org/hyperskill/app/study_plan/widget/presentation/StudyPlanWidgetReducer.kt]: I really wanted to do it on backend (pass `is_visible = false` for sections above), but it wasn't easy in realization, sorry(

**Review by vladkash — APPROVED** [review]

**Pull request merged**